File: apps/android/app/src/main/kotlin/com/swab/android/sync/SyncScheduler.kt
package com.swab.android.sync

import com.swab.android.observability.NoopLogger
import com.swab.android.observability.SwabLogger
import java.util.concurrent.atomic.AtomicLong
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Work waiting to leave the device. Today the only implementation is
 * `VaultSync` (one whole-blob push); after the ADR-001 stage-4 migration it
 * becomes the durable per-record outbox (VLT-10). The scheduler is written
 * against THIS, never against a vault or an ApiClient, so the triggers below
 * survive that swap untouched.
 */
interface PendingSync {
    suspend fun flush()
}

/**
 * FS-07 VLT-10 — *when* pending local work is replayed, decoupled from *what*
 * that work is (SUG-AND-001).
 *
 * **What the spec actually requires.** VLT-10: "Offline writes queue in a
 * durable local outbox and replay **in order** on reconnect." FS-01
 * acceptance 1: placements made in airplane mode "sync to the server when
 * connectivity returns, with no data loss". Before this class the app pushed
 * from exactly one place — the onboarding Done screen — so every fiche edit
 * made afterwards stayed on the device forever, and a single failed
 * onboarding push was never retried.
 *
 * **The specific triggers below are engineering choices, not requirements.**
 * FS-07's VLT-04 named "app background, post-onboarding, after any vault
 * write burst (debounced ≥30s)" until commit `ab3f241` (2026-08-16, ADR-001)
 * replaced it with the local-cache wording it carries now. The current spec
 * names no trigger and no interval. Post-onboarding, background, and a 30 s
 * debounce are this client's approximation of "on reconnect" while it has no
 * reachability callback — change them freely if a better approximation
 * appears; only the VLT-10 guarantee is binding.
 *  - **after a write burst**, debounced ([DEFAULT_DEBOUNCE_MILLIS]) — the
 *    fiche persists on every chip tap, so an editing session must collapse
 *    into one push, not one per tap;
 *  - **app background** ([onAppBackground]) — the debounce is cancelled and
 *    queued work goes out now, because the process may not survive to see the
 *    timer fire;
 *  - **post-onboarding** ([syncNow]) — unconditional: the server may hold no
 *    vault at all yet.
 *
 * Plus [onAppForeground], a poor-man's reconnect trigger: a push that failed
 * offline stays queued and is retried the next time the app comes back to the
 * front. (A real `ConnectivityManager` callback is deliberately NOT wired —
 * see the changelog; it needs a lifecycle-aware registration and belongs with
 * the stage-4 outbox.)
 *
 * **This is not the durable outbox VLT-10 asks for.** The queue counters live
 * in memory; durability arrives with ADR-001 stage 4. [assumePendingFromPreviousSession]
 * is the honest stopgap.
 *
 * No Android imports: this is plain-JVM testable with virtual time, like the
 * rest of the domain layer.
 *
 * Failures are silent by design (product ethos: nothing alarming). They are
 * logged at WARN with the exception *type* only (G3) and re-queued.
 */
class SyncScheduler(
    private val pending: PendingSync,
    /**
     * Must outlive any ViewModel — a debounce started by a fiche edit has to
     * survive leaving the fiche. `AppContainer` owns it.
     */
    private val scope: CoroutineScope,
    private val debounceMillis: Long = DEFAULT_DEBOUNCE_MILLIS,
    /**
     * ONB-05 gate. Nothing may leave the device before onboarding completes —
     * the privacy-invariant tests assert zero classification data on the wire
     * during onboarding, and that is current, correct behaviour. Writes made
     * during onboarding are *queued*, not dropped: the post-onboarding
     * [syncNow] flushes them.
     */
    private val isEnabled: suspend () -> Boolean = { true },
    private val logger: SwabLogger = NoopLogger(),
    /**
     * Monotonic clock for the backoff window. `nanoTime`, not
     * `currentTimeMillis`: a wall-clock jump must not turn a one-minute
     * cool-off into an hour (or vice versa). Tests inject virtual time.
     */
    private val nowMillis: () -> Long = { System.nanoTime() / 1_000_000 },
) {
    companion object {
        /**
         * Debounce window. An engineering choice, not a spec constant — see
         * the header: the current FS-07 names no interval.
         */
        const val DEFAULT_DEBOUNCE_MILLIS: Long = 30_000L

        /**
         * Backoff for a push that keeps failing the same way. One free retry
         * (a first failure is very often transient — a tunnel, a locked
         * Keystore), then doubling from a minute to a one-hour cap.
         *
         * Without this, a permanently-failing push is retried on EVERY
         * trigger, and issue #127 means the first push of every new account
         * can never succeed: with `onStart` and `onStop` both wired, each
         * app switch cost six doomed requests, forever. Failures are still
         * reported (G3) — they are just not re-attempted on a hair trigger.
         * Mirrors the iOS scheduler's values so the two platforms behave
         * alike for the same requirement.
         */
        const val BACKOFF_BASE_MILLIS: Long = 60_000L
        const val BACKOFF_CAP_MILLIS: Long = 3_600_000L

        internal fun backoffDelayMillis(consecutiveFailures: Int): Long {
            if (consecutiveFailures < 2) return 0L
            val doublings = minOf(consecutiveFailures - 2, 16)
            return minOf(BACKOFF_BASE_MILLIS shl doublings, BACKOFF_CAP_MILLIS)
        }
    }

    /** Serialises flushes so two triggers can't push the same state twice concurrently. */
    private val flushMutex = Mutex()

    /** Guards [debounceTimerJob] only — triggers are called from the main thread and from vault writes on other dispatchers. */
    private val jobLock = Any()
    private var debounceTimerJob: Job? = null

    /**
     * Local writes so far, and the count the server has accepted. They differ
     * exactly when something is queued.
     *
     * A boolean "dirty" flag is NOT enough and was the first version of this
     * class: a write landing *during* an in-flight push was cleared by that
     * push's success, even though the push carried the older state. The next
     * backgrounding then cancelled the debounce and skipped the flush on that
     * stale flag, and the edit never left the device — the exact defect this
     * whole change exists to fix. Snapshotting the sequence before the push
     * and only crediting that snapshot afterwards closes it.
     */
    private val writeSeq = AtomicLong(0)
    private val syncedSeq = AtomicLong(0)

    /** Set when a push fails, so a retry happens even with no new writes. */
    @Volatile
    private var lastFlushFailed: Boolean = false

    /**
     * Consecutive failures of the SAME kind, and the instant before which no
     * further attempt is made. A different failure type resets the count — a
     * changed failure mode is new information, not more of the same.
     */
    @Volatile
    private var consecutiveFailures: Int = 0

    @Volatile
    private var lastFailureType: String? = null

    @Volatile
    private var retryNotBefore: Long? = null

    /** True while a repeatedly-failing push is cooling off. */
    val isBackingOff: Boolean
        get() = retryNotBefore?.let { nowMillis() < it } ?: false

    /** Is there local state the server hasn't accepted yet? (Also the retry latch.) */
    val hasPendingWork: Boolean get() = lastFlushFailed || writeSeq.get() != syncedSeq.get()

    /**
     * Assume something is queued (FS-01 acceptance 1). Called at process
     * start once the persisted step says onboarding is finished: a push that
     * a previous session never landed leaves no in-memory trace — `writeSeq`
     * and `lastFlushFailed` die with the process — so without this the
     * airplane-mode onboarding case never syncs on the next launch, because
     * both cross-session triggers guard on [hasPendingWork]. The cost of
     * being wrong is one redundant push per app session; the cost of the
     * opposite is silent data loss. Mirrors iOS `SyncScheduler.activate()`.
     *
     * NOT durability: the real fix is VLT-10's durable outbox (ADR-001
     * stage 4). This is the honest approximation available without one.
     */
    fun assumePendingFromPreviousSession() {
        writeSeq.incrementAndGet()
    }

    /** Write-burst trigger — call on every persisted local write. */
    fun onWrite() {
        writeSeq.incrementAndGet()
        synchronized(jobLock) {
            debounceTimerJob?.cancel()
            debounceTimerJob = scope.launch {
                delay(debounceMillis)
                // The flush is its own job on the SCOPE, not a child of this
                // timer: cancelling the timer must never cancel a push
                // already in flight. It used to, and an accepted server write
                // whose local `setVaultVersion` got cancelled left the client
                // 409-ing on every later push.
                scope.launch { flush() }
            }
        }
    }

    /** Post-onboarding trigger — flushes immediately, pending work or not. */
    fun syncNow() {
        cancelDebounceTimer()
        scope.launch { flush() }
    }

    /**
     * App-background trigger. Cancels the debounce and pushes what is
     * queued. Deliberately a no-op when nothing is queued: the server bumps
     * the stored version on every accepted write, so re-pushing identical
     * bytes on each backgrounding would churn versions for nothing.
     */
    fun onAppBackground() {
        if (!hasPendingWork || isBackingOff) return
        cancelDebounceTimer()
        scope.launch { flush() }
    }

    /**
     * Retry latch: a push that failed while offline goes out when the app
     * returns. Cancels the debounce timer too — without that, a successful
     * flush here left the still-armed timer to fire 30 s later and push
     * byte-identical state a second time.
     */
    fun onAppForeground() {
        if (!hasPendingWork || isBackingOff) return
        cancelDebounceTimer()
        scope.launch { flush() }
    }

    private fun cancelDebounceTimer() {
        synchronized(jobLock) {
            debounceTimerJob?.cancel()
            debounceTimerJob = null
        }
    }

    private suspend fun flush() {
        flushMutex.withLock {
            if (!isEnabled()) return // ONB-05 — stays queued, nothing on the wire
            if (isBackingOff) return // cooling off after repeated identical failures
            // Snapshot BEFORE the push: anything written while it is in flight
            // is not covered by it and must stay queued (see [writeSeq]).
            val coveredSeq = writeSeq.get()
            try {
                pending.flush()
                syncedSeq.set(coveredSeq)
                lastFlushFailed = false
                consecutiveFailures = 0
                lastFailureType = null
                retryNotBefore = null
            } catch (cancellation: CancellationException) {
                lastFlushFailed = true // an interrupted push did not land
                throw cancellation
            } catch (t: Throwable) {
                // Throwable, not Exception, and deliberately so: this replaced
                // a `runCatching` on the Done screen, which caught Throwable.
                // Narrowing it would have been a silent crash regression —
                // this runs in a container-owned scope with no supervisor
                // above it, so ANY escape reaches SwabApplication's default
                // handler, which rethrows, and the process dies. A background
                // push must never be able to take the app down. Cancellation
                // is re-thrown above, so it stays real cancellation.
                //
                // Offline is the normal case here, not an error state: keep it
                // queued, keep the UI silent, log the type only (G3 — never
                // the blob, a version, or anything from the vault).
                lastFlushFailed = true
                val type = t.javaClass.simpleName
                consecutiveFailures = if (type == lastFailureType) consecutiveFailures + 1 else 1
                lastFailureType = type
                val cooloff = backoffDelayMillis(consecutiveFailures)
                retryNotBefore = if (cooloff == 0L) null else nowMillis() + cooloff
                logger.event(
                    SwabLogger.Level.WARN,
                    "sync.flush.failed",
                    mapOf("type" to type, "consecutive" to consecutiveFailures),
                )
            }
        }
    }
}

File: apps/ios/Sources/SwabCore/Sync/SyncScheduler.swift
/// FS-07 VLT-10 — *when* pending local work is replayed, decoupled from
/// *what* that work is.
///
/// **What the spec actually requires.** VLT-10: "Offline writes queue in a
/// durable local outbox and replay **in order** on reconnect." FS-01
/// acceptance 1: placements made in airplane mode "sync to the server when
/// connectivity returns (VLT-04), with no data loss". Before this type
/// existed the iOS app pushed from exactly one place —
/// `DoneViewModel.finish()` — so every fiche edit made after onboarding
/// stayed on the device forever, and a single failed onboarding push was
/// never retried. Offline onboarding completion is a first-class path, which
/// made that one push the one most likely to fail.
///
/// **The specific triggers below are engineering choices, not requirements.**
/// FS-07's VLT-04 named "app background, post-onboarding, after any write
/// burst (debounced ≥30s)" until commit `ab3f241` (2026-08-16, ADR-001)
/// replaced it with the local-cache wording it carries now. Nothing in the
/// current spec names a trigger or an interval. Post-onboarding, background,
/// and a 30 s debounce are this client's approximation of "on reconnect"
/// while it has no reachability callback — change them freely if a better
/// approximation appears; only the VLT-10 guarantee is binding.
///
/// **This is not the durable outbox VLT-10 asks for.** `needsSync` lives in
/// memory, so a session killed between a failed push and the next trigger
/// loses the retry. Durability arrives with ADR-001 stage 4.
///
/// **Deliberately not coupled to `VaultSync`.** ADR-001 stage 4 replaces the
/// vault blob with a local cache plus a durable offline outbox (VLT-10). The
/// *triggers* survive that swap; the *work* does not. So the scheduler drives
/// a `PendingSyncWork` — `VaultSync` conforms today, the outbox conforms
/// tomorrow — and nothing in this file mentions blobs, versions or HTTP.
import Foundation

/// One unit of "there is local state the server has not confirmed".
///
/// Throwing from `flushPendingWork()` means *still pending*: the scheduler
/// records the failure, keeps `needsSync` true, and retries on the next
/// trigger. It never retries in a loop of its own — a device that is offline
/// or locked must not be hammered.
public protocol PendingSyncWork: Sendable {
    func flushPendingWork() async throws

    /// G3: a fixed, privacy-safe code for a `flushPendingWork()` failure.
    /// The scheduler never touches `Error.localizedDescription` — a decode
    /// failure's context string can embed decoded payload fragments. The
    /// work owns the mapping because the work owns the error vocabulary.
    func reportCode(for error: Error) -> String
}

extension PendingSyncWork {
    public func reportCode(for error: Error) -> String { "syncFailed" }
}

public actor SyncScheduler {
    /// Debounce window. An engineering choice, not a spec constant — see the
    /// header: the current FS-07 names no interval.
    public static let debounceInterval: TimeInterval = 30

    public typealias Clock = @Sendable () -> Date
    public typealias Sleeper = @Sendable (TimeInterval) async throws -> Void

    /// Default sleeper. Injected rather than called directly so tests drive a
    /// deterministic clock instead of waiting out a real 30-second window.
    public static let systemSleep: Sleeper = { interval in
        guard interval > 0 else { return }
        try await Task.sleep(nanoseconds: UInt64(interval * 1_000_000_000))
    }

    /// Backoff for a push that keeps failing the same way.
    ///
    /// One free retry (the first failure is very often transient — a tunnel,
    /// a locked Keychain), then doubling from a minute up to an hour. Without
    /// this a permanently-failing push retries on *every* trigger: while
    /// issue #127 stands, the first push of every new account can never
    /// succeed, so an evening of five edits and ten backgroundings would be
    /// ~45 doomed requests. Failures are still reported (G3) — they are just
    /// not re-attempted on a hair trigger.
    public static let backoffBase: TimeInterval = 60
    public static let backoffCap: TimeInterval = 3600

    static func backoffDelay(afterConsecutiveFailures count: Int) -> TimeInterval {
        guard count >= 2 else { return 0 }
        let doublings = min(count - 2, 16)
        return min(backoffBase * pow(2, Double(doublings)), backoffCap)
    }

    private let work: PendingSyncWork
    private let reporter: ErrorReporter
    private let now: Clock
    private let sleep: Sleeper

    /// ONB-05 gate. Every trigger is inert until `activate()`: nothing may
    /// leave the device while onboarding's local-only window is open.
    private var isActive = false
    private var lastWriteAt: Date?
    private var debounceTask: Task<Void, Never>?
    private var isFlushing = false

    /// Bumped by `activate()` and by every `noteLocalWrite()`; a flush
    /// records the generation it *attempted*, so a write that lands while a
    /// flush is in flight is not swallowed by that flush's success.
    private var writeGeneration = 0
    private var flushedGeneration = 0
    private var lastFlushFailed = false

    /// A trigger that arrived while a flush was already running. The
    /// in-flight flush re-runs for it rather than dropping it (see
    /// `syncNow()`).
    private var flushRequestedDuringFlush = false

    /// Consecutive failures with the *same* report code, and the instant
    /// before which no further attempt is made. A different code resets the
    /// count — a changed failure is new information, not more of the same.
    private var consecutiveIdenticalFailures = 0
    private var retryNotBefore: Date?

    /// Fixed privacy-safe code of the last failure (G3), or nil after a
    /// confirmed push. Never a raw error string.
    public private(set) var lastError: String?

    /// True while the server has not confirmed the current local state.
    public var needsSync: Bool {
        writeGeneration != flushedGeneration || lastFlushFailed
    }

    /// True while a repeatedly-failing push is cooling off.
    public var isBackingOff: Bool {
        guard let retryNotBefore else { return false }
        return now() < retryNotBefore
    }

    public init(
        work: PendingSyncWork,
        reporter: ErrorReporter = NoopErrorReporter(),
        now: @escaping Clock = { Date() },
        sleep: @escaping Sleeper = SyncScheduler.systemSleep
    ) {
        self.work = work
        self.reporter = reporter
        self.now = now
        self.sleep = sleep
    }

    // MARK: - Triggers

    /// ONB-05's window has closed — arm the triggers.
    ///
    /// Marks work pending without pushing: whatever was written before this
    /// point (onboarding's contacts and calibration, or a previous session
    /// that died before its push landed) has never been confirmed by the
    /// server, and nothing durable records it yet — the outbox is VLT-10 /
    /// ADR-001 stage 4. The cost of being wrong is one redundant push per
    /// app session; the cost of the opposite is silent data loss.
    public func activate() {
        guard !isActive else { return }
        isActive = true
        writeGeneration += 1
    }

    /// Trigger 1 — post-onboarding.
    public func onboardingDidComplete() async {
        activate()
        await syncNow()
    }

    /// Trigger 2 — a local write. Debounced: a burst of edits
    /// coalesces into one push, `debounceInterval` after the last of them.
    public func noteLocalWrite() {
        guard isActive else { return }
        writeGeneration += 1
        lastWriteAt = now()
        startDebounceIfNeeded()
    }

    /// Trigger 3 — the app went to the background. Cancels the
    /// pending debounce (waiting it out in the background is exactly what
    /// there is no time for) and pushes immediately, if anything is pending.
    public func appDidEnterBackground() async {
        guard isActive else { return }
        cancelDebounce()
        guard needsSync else { return }
        await syncNow()
    }

    /// Flush whatever is pending. Never throws: a failed sync is a normal
    /// state, not an error the caller has to handle.
    public func syncNow() async {
        guard isActive else { return }
        guard !isBackingOff else { return }
        // Actors are reentrant: a trigger can land while a flush is already
        // in flight. Record it rather than dropping it — the running flush
        // re-runs below, so a write that arrives mid-push is never stranded
        // with nothing scheduled to carry it.
        guard !isFlushing else {
            flushRequestedDuringFlush = true
            return
        }
        isFlushing = true
        repeat {
            flushRequestedDuringFlush = false
            await performFlush()
        } while flushRequestedDuringFlush && needsSync && !isBackingOff
        isFlushing = false
    }

    private func performFlush() async {
        // Captured BEFORE the await: a write landing mid-flush bumps
        // `writeGeneration` past `attempted`, so this flush cannot claim
        // work it did not carry.
        let attempted = writeGeneration
        do {
            try await work.flushPendingWork()
            flushedGeneration = attempted
            lastFlushFailed = false
            lastError = nil
            consecutiveIdenticalFailures = 0
            retryNotBefore = nil
        } catch {
            // Retryable by construction. The Keychain that holds the session
            // token is `kSecAttrAccessibleWhenUnlockedThisDeviceOnly`
            // (`Identity/SecureStore.swift`), so a background sync running
            // after the device locks legitimately fails to read it; a
            // backgrounded push can also be cut short by suspension. Neither
            // is fatal — `needsSync` stays true and a later trigger retries.
            lastFlushFailed = true
            let code = work.reportCode(for: error)
            consecutiveIdenticalFailures = (code == lastError) ? consecutiveIdenticalFailures + 1 : 1
            lastError = code
            let delay = Self.backoffDelay(afterConsecutiveFailures: consecutiveIdenticalFailures)
            retryNotBefore = delay > 0 ? now().addingTimeInterval(delay) : nil
            // `domain` stays "vault.sync": it is the established Console.app
            // category for this failure (see `OSLogErrorReporter`), inherited
            // from `DoneViewModel`. A stable log category, not a coupling —
            // nothing here knows what a vault is.
            reporter.report(
                ReportedError(domain: "vault.sync", operation: "sync", errorDescription: code)
            )
        }
    }

    // MARK: - Debounce

    /// At most one debounce task exists at a time. A new write does not spawn
    /// a second one — it moves `lastWriteAt`, and the running task re-reads
    /// the deadline on every wake. (Plan gotcha: "cancelled/restarted per
    /// change, not accumulated, or rapid edits leak tasks".)
    private var isDebouncing: Bool {
        guard let debounceTask else { return false }
        return !debounceTask.isCancelled
    }

    private func startDebounceIfNeeded() {
        guard !isDebouncing else { return }
        let sleeper = sleep
        let clock = now
        debounceTask = Task { [weak self] in
            while !Task.isCancelled {
                guard let remaining = await self?.remainingDebounce(at: clock()) else { return }
                if remaining <= 0 { break }
                do {
                    try await sleeper(remaining)
                } catch {
                    return  // cancelled
                }
            }
            guard !Task.isCancelled else { return }
            await self?.debounceDidFire()
        }
    }

    private func remainingDebounce(at instant: Date) -> TimeInterval {
        guard let lastWriteAt else { return 0 }
        return lastWriteAt.addingTimeInterval(Self.debounceInterval).timeIntervalSince(instant)
    }

    private func debounceDidFire() async {
        debounceTask = nil
        // Same guard as the background trigger: if a push already carried
        // these writes (a backgrounding beat the debounce), there is nothing
        // to send.
        guard needsSync else { return }
        await syncNow()
    }

    private func cancelDebounce() {
        debounceTask?.cancel()
    }

    // MARK: - Test seams (`@testable` only — not part of the public surface)

    /// Awaits the in-flight debounce task, including the sync it fires.
    func drainDebounceForTests() async {
        for _ in 0..<8 {
            guard let task = debounceTask else { return }
            await task.value
            if debounceTask == task {
                debounceTask = nil
                return
            }
        }
    }

    var debounceTaskCountForTests: Int { debounceTask == nil ? 0 : 1 }

    var writeGenerationForTests: Int { writeGeneration }
}
